### PR TITLE
FrontEnd: Orders Summary Component And UserProfilePage

### LINKED ISSUE #213: UserProfilePage /account Info Container
### User story
Who: As a developer
What: I want to develop the UserProfilePage 
Why: So that I can display the user profile page/ Account page.

### Acceptance criteria
This page should look according to screenshot below:
Screen Shot 2022-03-02 at 11 10 18 PM

### Implementation details
You will need  Account Page container  with all the information of user and their order summary .

Use the container and style file mentioned on this https://github.com/HackYourFuture-CPH/fp-class19/issues/32

### LINKED ISSUE #224: FrontEnd: Order Summary Component
## User story ## 
Who: As a developer
What: I want to develop the Orders Summary Component
Why: So that it will display the orders summary of the user.

## Acceptance criteria ##
This page should look according to screenshot below:
Screenshot 2022-03-05 at 19 00 36
Check the wireframes here:
https://www.figma.com/file/yo4QT2EcPo83HAMI3ydMxb/HYF-Wireframes?node-id=216%3A28468

## Implementation details ##
Create a form which takes user credentials

### PR DESCRIPTION
**Pull request by suvarna-ratna**
# Description
This PR displays orders summary in the userProfilePage
Please provide a short summary explaining what this PR is about.

Fixes #224
Fixes #213 

# How to test?
npm run storybook
Please provide a short summary how your changes can be tested?

# Checklist

- [X] I have performed a self-review of my own code
- [X] I have followed the name conventions for CSS Classnames and filenames, Components names and filenames, Style filenames, if you are in doubt check the the project README.MD and here https://github.com/HackYourFuture-CPH/curriculum/blob/master/review/review-checklist.md
- [X] I have commented my code, particularly in hard-to-understand areas, if you code was simple enough mark the box anyway
- [X] I have made corresponding changes to the documentation, if you code was simple enough mark the box anyway
- [X] This PR is ready to be merged and not breaking any other functionality

### CODE REVIEW

**Review by saloumeh-67 — CHANGES_REQUESTED** [review]
it works , but the look is not the same
Screen Shot 2022-03-06 at 3 39 45 PM
 as wireframe! each summery is in  individual box , but in your storybook they all in one box.  and colors should be the same as wireframe too :)

**Comment by Impressiv3** [comment]
I'm trying to test it but at the moment I can only test it in storybook and it looks like this:
image
Am I missing something here ? :D

File: src/client/components/OrdersSummary/OrdersSummary.stories.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+OrdersDisplay.args = {
+  orders: [
+    {
+      id: 1,
+      user_id: 1,
+      status: 'NEW',
+      created_at: "2022-02-28 21:28:04",
+      updated_at: "2022-02-28 21:28:04",
+    },
+    {
```
Comment (FlorentinaPetica) [inline src/client/components/OrdersSummary/OrdersSummary.stories.js:26]: As Balint is saying we can see a storybook but the issue #224 is for the component. Is the issue wrong?

For order summary you will be fetching first this api `api/orders?user=user_id` and the response you are going to receive is looking like this: 
image

And after this when the user clicks on an order we will show the order dispaly which will call the api `api/orders/{id}`  which returns something like what you have here.
image

**Review by FlorentinaPetica — CHANGES_REQUESTED** [review]

**Comment by suvarna-ratna** [comment]
> it works , but the look is not the same Screen Shot 2022-03-06 at 3 39 45 PM as wireframe! each summery is in individual box , but in your storybook they all in one box. and colors should be the same as wireframe too :)

Fixed now .. Pls have a look once .. Thank you

File: src/client/components/OrdersSummary/OrdersSummary.stories.js (hunk 1 above)
Comment (suvarna-ratna) [reply src/client/components/OrdersSummary/OrdersSummary.stories.js:26]: Thank you for Info Florentina.. Please have a look once and give feedback please ..

**Comment by suvarna-ratna** [comment]
> I'm trying to test it but at the moment I can only test it in storybook and it looks like this: image Am I missing something here ? :D

Fixed now .. Thank you for feedback

File: src/client/components/OrdersSummary/OrderComponent.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            </h3>
+          </div>
+          <div>
+            <p>{moment(order.updated_at).format('DD-MM-YYYY')}</p>
+            <p>Kr. {order.total_amount}</p>
+          </div>
+          <div>
+            {' '}
+            <button type="button" onClick={viewOrder}>
+              {fullView ? <p>&#x25BC;</p> : <p>&#x25B2;</p>}
```
Comment (FlorentinaPetica) [inline src/client/components/OrdersSummary/OrderComponent.js:29]: I whould swap this arrows they look in the wrong order. But for the rest the storybook is working 👍

**Review by FlorentinaPetica — APPROVED** [review]

File: src/client/components/OrdersSummary/OrdersSummary.styles.css [hunk 3]
```diff
@@ -0,0 +1,32 @@
+.orders-view .order-item {
+  border: solid white 3px;
+  background-color: #f6f8eb;
+  margin-bottom: 15px;
+}
+
+.orders-view button {
```
Comment (miltonhultgren) [inline src/client/components/OrdersSummary/OrdersSummary.styles.css:7]: Nitpick:
Avoid naming your CSS classes after which HTML tag is being used, you might change the tag but the class name should not need to change for that.

Similarly, don't name the class based on what CSS properties it uses, like "flex-container".

Aim instead to name your CSS classes for what the objects on the page mean to the user!

File: src/client/components/OrdersSummary/OrdersSummary.component.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import OrderComponent from './OrderComponent';
+import './OrdersSummary.styles.css';
+
+export default function OrdersSummary({ orders, orderDetails }) {
```
Comment (miltonhultgren) [inline src/client/components/OrdersSummary/OrdersSummary.component.js:5]: This is fairly inefficient. I imagine it's done this way because it's just for Storybook in this PR and you might not know yet where to put the data or call the API but this code won't perform very well.
Let's see how we can make it better by understanding what it does poorly.

On each render, you will loop over `orders`, and each time you render one `OrderComponent` you will filter the list of `orderDetails` to find the relevant ones. 
Imagine you have 100 orders, and each one has 10 items, `orderDetails` will be 1000 items. Now, that isn't so much but say that your app re-renders often? On top of this, what if the user never clicks to view the details of an order? We're doing all this work for no reason.

Imagine instead, if you had an interface like this:
```js
function OrderSummaries({ orders, loadDetailsFor }) {
  return (
    <div className="orders-view">
        <ul className="order-list">
          {orders.map((order) => (
            <li className="order-item">
              <OrderSummary order={order} loadDetailsFor={loadDetailsFor} />
            </li>
          ))}
        </ul>
    </div>
  );
}

function OrderSummary({ order, loadDetailsFor }) {
  const [showDetails, setShowDetails] = useState(false);
  const [orderDetails, setOrderDetails] = useState(null);

  const onToggleDetails = useCallback(() => {
    const shouldShowDetails = !showDetails;
    setShowDetails(shouldShowDetails );

    if (shouldShowDetails && !orderDetails) {
      loadDetailsFor(order.id)
        .then((details) => setOrderDetails(details));
    }
  }, [loadDetailsFor]);

  // As a bonus, we could extract smaller components here.
  return (
    <OrderSummaryRow order={order} onToggleDetails={onToggleDetails} />
    {showDetails && orderDetails &&
      <OrderSummaryDetails details={orderDetails} />
    }
  );
}
```

No more looping on each render and we only load the data we need when we need it.
But where does that logic live then? In the rest of the app, like the Page container.

```js
function OrderSummariesPage({ user_id }) {
  const orders = useOrders(user_id); // Fetches orders for the user from the API, without the details

  const loadDetailsFor = (orderId) => {
    return fetch(`api/orders/${orderId}`); // Fetch returns a Promise for the data
  };

  return (
    <OrderSummaries orders={orders} loadDetailsFor={loadDetailsFor} />
  );
}
```

This way we only load the details when we need to and only once. Note also that the `OrderSummary` component doesn't actually know how to load that data, it just knows how to tell the app that now is when that data is needed.

The big benefits is that you separate components that know about the data from those that don't, as much as possible. And only load and pass around the data that is actually needed to render right now.

**Review by miltonhultgren — CHANGES_REQUESTED** [review]
I left a long comment on how we could change the flow of data in this code to make it easier to manage once you actually put the APIs in. The current code won't handle real data well.

Don't hesitate to reach out on Slack if anything is unclear, for this PR it's likely enough to change only the components and the Storybook, the sample code for how the page should work is out of scope and just to highlight how it all fits together.

File: src/client/containers/UserProfilePage/UserProfilePage.Container.js [hunk 5]
```diff
@@ -1,11 +1,23 @@
...
+  const orderDetails = (userId) => fetch(`api/orders/${userId}`);
+
+  if (isLoading) {
+    return <Loader />;
+  }
+  
   return (
     <div>
       <UserProfile />
+      <OrdersSummary orders={orders(user.uid)} orderDetails={orderDetails(user.uid)}/>
```
Comment (miltonhultgren) [inline src/client/containers/UserProfilePage/UserProfilePage.Container.js:20]: I don't see how this would work.
Both `orders` and `orderDetails` return Promises. There is no code here that either adds a `.then` handler or uses `await` to resolve those Promises. Later you're calling `.map` on a Promise which won't work.

File: src/server/api/controllers/orders.controller.js [hunk 6]
```diff
@@ -112,7 +112,8 @@ const getOrdersByUserId = async (raw_user_id) => {
       .groupBy('orders.id')
       .count('orders.id as nr_of_items');
     if (orders.length === 0) {
-      throw new HttpError('The user have no orders yet', 404);
+      return [];
+      // throw new HttpError('The user have no orders yet', 404);
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js:116]: Why was this changed?

File: src/client/components/OrdersSummary/OrdersSummary.component.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  
+  const filteredDetails = (id) => orderDetails.filter((odr) => odr.id === id);
+  return (
+    <div className="orders-view">
+      <div>
+        <ul className="order-list">
+          {orders.map((each) => (
+            <li className="order-item">
+              <OrderComponent order={each} details={filteredDetails(each.order_number)}/>
```
Comment (miltonhultgren) [inline src/client/components/OrdersSummary/OrdersSummary.component.js:14]: This still does the unneeded looping.

**Review by miltonhultgren — COMMENTED** [review]
@dpfernandes I don't know what to do about this PR, I understand time is running out but this PR just doesn't seem merge ready.

File: src/client/components/OrdersSummary/OrdersSummary.component.js (hunk 4 above)
Comment (suvarna-ratna) [reply src/client/components/OrdersSummary/OrdersSummary.component.js:5]: Thank you changes done plz have a look on it .

File: src/client/containers/UserProfilePage/UserProfilePage.Container.js (hunk 5 above)
Comment (suvarna-ratna) [reply src/client/containers/UserProfilePage/UserProfilePage.Container.js:20]: Hi Milton i tried a lot ,i will see after merge this, how it actually work ... i will fix changes later .
